### PR TITLE
Support coverage v0.9.2

### PR DESCRIPTION
**Pull request by nilsdoehring**
Hi @Adracus, as discussed elsewhere, here are the changes required for dart-coveralls to work again. Note: In `collect_lcov.dart` L114, the process doesn't complete the `Future`, thus I refactored to not use `await`. It will do for now, but with the changes to Observatory URLs coming with Dart 1.22.0, it will require a revisit.

### CODE REVIEW

**Comment by matanlurey** [comment]
FWIW see https://github.com/duse-io/dart-coveralls/issues/53

`package:package_config` has some logic to handle both .packages and packages/, and whatever changes occur in the future without needing to change dart-coveralls - https://github.com/dart-lang/package_config/blob/master/lib/discovery.dart#L93.

File: bin/src/upload.dart [hunk 1]
```diff
@@ -27,7 +27,7 @@ class UploadPart extends CommandLinePart {
       });
     }
 
-    var pRoot = new Directory(res["package-root"]);
+    var pRoot = new File(res["packages"]);
```
Comment (Adracus) [inline bin/src/upload.dart]: Can we provide some way that is also backward compatible? As of now, every script that used `package-root` would break. I'd rather still support it but provide deprecation warnings.

File: pubspec.yaml [hunk 2]
```diff
@@ -1,15 +1,15 @@
...
-version: 0.4.0
+version: 0.5.0
 author: Axel Christ <adracus@gmail.com>
 description: |-
   Pub package to calculate coverage, format it
   to LCOV and send it to coveralls
 homepage: https://github.com/duse-io/dart-coveralls
 environment:
-  sdk: '>=1.9.0 <2.0.0'
+  sdk: '>=1.9.0 <1.22.0'
```
Comment (Adracus) [inline pubspec.yaml]: I've not been in the dartland for a long time - It seems the dev branch is currently failing when this is restricted to 1.22 - Should this be adjusted or what is the reason for this?

File: bin/src/report.dart [hunk 3]
```diff
@@ -42,9 +42,9 @@ class ReportPart extends CommandLinePart {
 
     if (res.rest.length != 1) return print("Please specify a test file to run");
 
-    var pRoot = new Directory(res["package-root"]);
-    if (!pRoot.existsSync()) return print("Root directory does not exist");
-    log.info(() => "Package root is ${pRoot.absolute.path}");
+    FileSystemEntity pRoot = new File(res["packages"]);
```
Comment (Adracus) [inline bin/src/report.dart]: Should we maybe, as mentioned by @matanlurey, use the package api for this now?
Again for reference:
https://github.com/dart-lang/package_config/blob/master/lib/discovery.dart#L93

**Comment by eseidel** [comment]
https://github.com/eseidel/dart-coveralls/commit/83ddd50d30319be59dcae992d048e13c9b8e7127 has an alternative implementation of some of this in case it's useful.  I'm happy to have yours land instead, since mine only does the .package bits.

File: bin/src/calc.dart [hunk 4]
```diff
@@ -60,6 +60,9 @@ ArgParser _initializeParser() => new ArgParser(allowTrailingOptions: true)
...
   ..addOption("workers", help: "Number of workers for parsing", defaultsTo: "1")
   ..addOption("output", help: "Output file path")
+  ..addOption("packages",
+      help: 'Where to find the packages file, that is, "package:..." imports.',
+      defaultsTo: ".packages")
   ..addOption("package-root",
-      help: 'Where to find packages, that is, "package:..." imports.',
-      defaultsTo: "packages");
+      help: 'Ignored/Deprecated. Package directories are no longer supported.',
+      defaultsTo: ".packages");
```
Comment (eseidel) [inline bin/src/calc.dart]: I don't think you meant to change 'packages' to '.packages' on this line.

File: bin/src/upload.dart (hunk 1 above)
Comment (eseidel) [inline bin/src/upload.dart]: The SDK restriction should stop that in part?  Clients who want to upgrade past dart 1.20 are either going to have to change to use pub get --packages-dir, or will need to update their code to account for .packages anyway?  At least that was my logic when I wrote my version of this.

File: bin/src/upload.dart [hunk 5]
```diff
@@ -77,9 +77,12 @@ ArgParser _initializeParser() => new ArgParser(allowTrailingOptions: true)
...
       help: "Token for coveralls", defaultsTo: Platform.environment["test"])
   ..addOption("workers", help: "Number of workers for parsing", defaultsTo: "1")
+  ..addOption("packages",
+      help: 'Where to find the packages file, that is, "package:..." imports.',
+      defaultsTo: ".packages")
   ..addOption("package-root",
-      help: 'Where to find packages, that is, "package:..." imports.',
-      defaultsTo: "packages")
+      help: 'Ignored/Deprecated. Package directories are no longer supported.',
+      defaultsTo: ".packages")
```
Comment (eseidel) [inline bin/src/upload.dart]: Same spurious rename as before.  I think you want to leave this line as 'packages' not '.packages', not that it matters hugely since its ignored.

File: lib/src/collect_lcov.dart [hunk 6]
```diff
@@ -102,12 +102,43 @@ class LcovCollector {
...
+        stderr.writeln('stderr:');
+        stderr.writeln(result.stderr);
+        throw new ProcessException(Platform.executable, dartArgs,
+            'There was a critical error. Exit code: ${result.exitCode}',
+            result.exitCode);
+      }
+    });
+
+    var cvArgs = [
+      "global",
```
Comment (eseidel) [inline lib/src/collect_lcov.dart]: I'm not sure you want to modify the users global pubcache.

File: bin/src/upload.dart (hunk 1 above)
Comment (nilsdoehring) [inline bin/src/upload.dart]: @Adracus @eseidel my focus was on getting `dart_coveralls` to run at all, as it currently only works with `coverage` versions prior to v0.7.4 (only `.packages` file allowed as input), which in turn will only work with SDK versions prior to v1.16.0.

Accordingly, backwards compatibility would have to introduce checks on SDK version and `coverage` package version, and combinations thereof, additionally depending on whether a file or dir was given as package source. Looks like quite some effort to get this right, especially since the `.packages` file is the way to go since v1.12.0. 

TL;DR: support for `package-root` dir requires an old version of `coverage` that doesn't work with Dart SDK >= 1.16.0. I don't think there's demand for this.

File: pubspec.yaml (hunk 2 above)
Comment (nilsdoehring) [inline pubspec.yaml]: The reason is a [breaking change](https://groups.google.com/a/dartlang.org/forum/#!topic/announce/VxSw-V5tx8k) to the VM service protocol in SDK 1.22.0, that I didn't account for with this PR. I have revisited the case and now got it running on the 1.22 dev branch.

TL;DR: underlying issue fixed, will change back to <2.0.0

File: bin/src/report.dart (hunk 3 above)
Comment (nilsdoehring) [inline bin/src/report.dart]: @Adracus @matanlurey thanks for the pointer to that package. Thing is, as explained in my comment on `upload.dart`, that the `coverage` package exclusively supports an option to provide a package file since v0.7.4 (i.e. package directories not supported anymore), and I don't see use in providing that-far-backwards compatibility.

File: bin/src/calc.dart (hunk 4 above)
Comment (nilsdoehring) [inline bin/src/calc.dart]: That might be correct. I'll also see to provide a better explanation text. I only left this in to have old implementations not break on the args level – they currently break script-wise, but at least silently, i.e. not bothering travis. Never change a (silently) failing system ;-)

File: bin/src/upload.dart (hunk 5 above)
Comment (nilsdoehring) [inline bin/src/upload.dart]: Yup. See other spurious comment.

File: lib/src/collect_lcov.dart (hunk 6 above)
Comment (nilsdoehring) [inline lib/src/collect_lcov.dart]: It was the easiest way to get to a result that `dart_coveralls` can consume without refactoring, but I'm halfway through of getting it to work via direct invocation of `coverage` API methods.

**Comment by nilsdoehring** [comment]
Thanks for the input, @eseidel @matanlurey @Adracus – next PR will:
* work only with SDK >1.16.0 (only to be revisited on reasonable demand)
* work also with SDK >=1.22.0 (observatory URI's coming with token)
* steal some implementation detail from [eseidel's PR](https://github.com/eseidel/dart-coveralls/commit/83ddd50d30319be59dcae992d048e13c9b8e7127)
* not use `pub global` to get results from `coverage` package
* be more clear about the `package-root` option being ignored
